diff --git python-package/pyproject.toml python-package/pyproject.toml
index 8835def25..c5dc908d9 100644
--- python-package/pyproject.toml
+++ python-package/pyproject.toml
@@ -6,7 +6,7 @@ backend-path = ["."]
 build-backend = "packager.pep517"
 
 [project]
-name = "xgboost"
+name = "xgboost-cpu"
 description = "XGBoost Python Package"
 readme = { file = "README.rst", content-type = "text/x-rst" }
 authors = [
@@ -30,8 +30,7 @@ classifiers = [
 ]
 dependencies = [
     "numpy",
-    "scipy",
-    "nvidia-nccl-cu12 ; platform_system == 'Linux' and platform_machine != 'aarch64'"
+    "scipy"
 ]
 
 [project.urls]
@@ -83,3 +82,6 @@ class-attribute-naming-style = "snake_case"
 
 # Allow single-letter variables
 variable-rgx = "[a-zA-Z_][a-z0-9_]{0,30}$"
+
+[tool.hatch.build.targets.wheel]
+packages = ["xgboost/"]
