diff --git python-package/pyproject.toml python-package/pyproject.toml
index 8835def25..42400a78a 100644
--- python-package/pyproject.toml
+++ python-package/pyproject.toml
@@ -6,7 +6,7 @@ backend-path = ["."]
 build-backend = "packager.pep517"
 
 [project]
-name = "xgboost"
+name = "xgboost-cpu"
 description = "XGBoost Python Package"
 readme = { file = "README.rst", content-type = "text/x-rst" }
 authors = [
@@ -30,8 +30,7 @@ classifiers = [
 ]
 dependencies = [
     "numpy",
-    "scipy",
-    "nvidia-nccl-cu12 ; platform_system == 'Linux' and platform_machine != 'aarch64'"
+    "scipy"
 ]
 
 [project.urls]
@@ -48,6 +47,10 @@ pyspark = ["pyspark", "scikit-learn", "cloudpickle"]
 
 [tool.hatch.build.targets.wheel.hooks.custom]
 
+
+[tool.hatch.build.targets.wheel]
+packages = ["xgboost/"]
+
 [tool.isort]
 profile = "black"
 
